File: communicationAPI/src/main/java/eu/cybergeiger/communication/PluginInformation.java
package eu.cybergeiger.communication;

import ch.fhnw.geiger.serialization.Serializer;
import ch.fhnw.geiger.serialization.SerializerHelper;
import ch.fhnw.geiger.totalcross.ByteArrayInputStream;
import ch.fhnw.geiger.totalcross.ByteArrayOutputStream;
import java.io.IOException;

/**
 * <p>Object for storing vital plugin information.</p>
 */
public class PluginInformation implements Serializer {

  private static final long serialVersionUID = 48032940912340L;

  private final String executable;
  private final int port;
  private CommunicationSecret secret;

  public PluginInformation(String executable, int port) {
    this(executable, port, null);
  }

  /**
   * <p>Constructor for plugin information.</p>
   *
   * @param executable the string required for platform specific wakeup of a plugin
   * @param port       the port of the plugin to be contacted on
   * @param secret     the secret required for communicating (if null a new secret is generated)
   */
  PluginInformation(String executable, int port, CommunicationSecret secret) {
    this.executable = executable;
    this.port = port;
    this.secret = secret;
    if (this.secret == null) {
      this.secret = new CommunicationSecret();
    }
  }

  /**
   * <p>Gets the port of the plugin.</p>
   *
   * @return the port an active plugin may be reached on
   */
  public int getPort() {
    return this.port;
  }

  /**
   * <p>The executable string required for starting the plugin.</p>
   *
   * @return the executable string
   */
  public String getExecutable() {
    return this.executable;
  }

  /**
   * <p>The communication secret required for sending securely between two instances.</p>
   *
   * @return the requested secret
   */
  public CommunicationSecret getSecret() {
    return this.secret;
  }

  @Override
  public void toByteArrayStream(ByteArrayOutputStream out) throws IOException {
    SerializerHelper.writeLong(out, serialVersionUID);
    SerializerHelper.writeString(out, executable);
    SerializerHelper.writeInt(out, port);
    secret.toByteArrayStream(out);
    SerializerHelper.writeLong(out, serialVersionUID);
  }

  /**
   * <p>Reads objects from ByteArrayInputStream and stores them in map.</p>
   *
   * @param in ByteArrayInputStream to be used
   * @return the deserialized Storable String
   * @throws IOException if value cannot be read
   */
  public static PluginInformation fromByteArrayStream(ByteArrayInputStream in) throws IOException {
    if (SerializerHelper.readLong(in) != serialVersionUID) {
      throw new ClassCastException();
    }

    String executable = SerializerHelper.readString(in);
    int port = SerializerHelper.readInt(in);
    CommunicationSecret secret = CommunicationSecret.fromByteArrayStream(in);

    if (SerializerHelper.readLong(in) != serialVersionUID) {
      throw new ClassCastException();
    }

    return new PluginInformation(executable, port, secret);

  }

  /**
   * <p>Wrapper function to simplify serialization.</p>
   *
   * @return the serializer object as byte array
   */
  public byte[] toByteArray() {
    try {
      ByteArrayOutputStream out = new ByteArrayOutputStream();
      toByteArrayStream(out);
      return out.toByteArray();
    } catch (IOException e) {
      return null;
    }
  }

  /**
   * <p>Wrapper function to simplify deserialization.</p>
   *
   * @param buf the buffer to be read
   * @return the deserialized object
   */
  public static PluginInformation fromByteArray(byte[] buf) {
    try {
      ByteArrayInputStream in = new ByteArrayInputStream(buf);
      return fromByteArrayStream(in);
    } catch (IOException ioe) {
      ioe.printStackTrace();
      return null;
    }
  }

  @Override
  public int hashCode() {
    return (executable + ":" + port + ":" + secret.toString()).hashCode();
  }

}

File: lib/src/communication/PluginInformation.dart
import 'CommunicationSecret.dart';

/// Object for storing vital plugin information.
class PluginInformation /*with ch_fhnw_geiger_serialization_Serializer*/ {
  static const int serialVersionUID = 48032940912340;
  final String? executable;
  final int port;
  late CommunicationSecret secret;

  /// Creates a [PluginInformation] with the given properties:
  ///
  /// - [executable] the string required for platform specific wakeup of a plugin
  /// - [port]       the port of the plugin to be contacted on
  /// - [secret]     the secret required for communicating (if null a new secret is generated)
  PluginInformation(this.executable, this.port, [CommunicationSecret? secret]) {
    this.secret = secret ?? CommunicationSecret.empty();
  }

  /// Gets the port of the plugin.
  int getPort() {
    return port;
  }

  /// The executable string required for starting the plugin.
  String? getExecutable() {
    return executable;
  }

  /// The communication secret required for sending securely between two instances.
  CommunicationSecret getSecret() {
    return secret;
  }

  /*void toByteArrayStream(ch_fhnw_geiger_totalcross_ByteArrayOutputStream out)
    {
        SerializerHelper.writeLong(out, serialVersionUID);
        SerializerHelper.writeString(out, executable);
        SerializerHelper.writeInt(out, port);
        secret.toByteArrayStream(out);
        SerializerHelper.writeLong(out, serialVersionUID);
    }

    /// Reads objects from ByteArrayInputStream and stores them in map.
    /// @param in ByteArrayInputStream to be used
    /// @return the deserialized Storable String
    /// @throws IOException if value cannot be read
    static PluginInformation fromByteArrayStream(ch_fhnw_geiger_totalcross_ByteArrayInputStream in_)
    {
        if (SerializerHelper.readLong(in_) != serialVersionUID) {
            throw new ClassCastException();
        }
        String executable = SerializerHelper.readString(in_);
        int port = SerializerHelper.readInt(in_);
        CommunicationSecret secret = CommunicationSecret_.fromByteArrayStream(in_);
        if (SerializerHelper.readLong(in_) != serialVersionUID) {
            throw new ClassCastException();
        }
        return new PluginInformation(executable, port, secret);
    }

    /// Wrapper function to simplify serialization.
    /// @return the serializer object as byte array
    List<int> toByteArray()
    {
        try {
            ch_fhnw_geiger_totalcross_ByteArrayOutputStream out = new ch_fhnw_geiger_totalcross_ByteArrayOutputStream();
            toByteArrayStream(out);
            return out.toByteArray();
        } on java_io_IOException catch (e) {
            return null;
        }
    }

    /// Wrapper function to simplify deserialization.
    /// @param buf the buffer to be read
    /// @return the deserialized object
    static PluginInformation fromByteArray(List<int> buf)
    {
        try {
            ch_fhnw_geiger_totalcross_ByteArrayInputStream in_ = new ch_fhnw_geiger_totalcross_ByteArrayInputStream(buf);
            return fromByteArrayStream(in_);
        } on java_io_IOException catch (ioe) {
            ioe.printStackTrace();
            return null;
        }
    }*/

  int get hashCode {
    return ((executable ?? 'null') +
            ':' +
            port.toString() +
            ':' +
            secret.toString())
        .hashCode;
  }
}
